src/library/searchquery.cpp

```diff
@@ -193,9 +215,37 @@ QString CrateFilterNode::toSql() const {
             m_pCrateStorage->formatQueryForTrackIdsByCrateNameLike(m_crateNameLike));
 }
 
+
+NoCrateFilterNode::NoCrateFilterNode(const CrateStorage* pCrateStorage)
+    : m_pCrateStorage(pCrateStorage),
+      m_matchInitialized(false) {
+}
+
+bool NoCrateFilterNode::match(const TrackPointer& pTrack) const {
+    if (!m_matchInitialized) {
+        TrackSelectResult tracks(
+                m_pCrateStorage->selectAllTracksSorted());
```

Comment:
I'm a bit worried, because generally the size of this query's result if of the same magnitude as the size of the whole library.

The only way for preventing this would be to use an adaptive strategy depending on the actual contents of the crate tracks table, i.e. for deciding if most tracks are organized in crates or not. This comes at the cost of one or two additional queries for determining the sizes upfront to chose the optimum query that will return either a blacklist or a whitelist.

But the whole search thing does not scale very well, so let's just keep it as is without introducing more complexity.